### ISSUE TITLE
 [FAIL] Kubevirt Virtual Machines when live migrated [It] with pre-copy should keep connectivity

### ISSUE THREAD

**Issue by trozet** [ovn-org/ovn-kubernetes#3986]
Seeing failures with this test still:

```
Kubevirt Virtual Machines when live migrated [It] with pre-copy should keep connectivity
/home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/kubevirt.go:656

  [FAILED] Oct 25 23:45:54.083: Timed out after 60.010s.
  worker1: after live migrate, migration #1: Check tcp connection is not broken
  Expected success, but got an error:
      <*errors.errorString | 0xc0003bb4a0>: 
      http connection to virtual machine was broken
      {
          s: "http connection to virtual machine was broken",
      }
  In [It] at: /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/kubevirt.go:296 @ 10/25/23 23:45:54.084

  There were additional failures detected.  To view them in detail run ginkgo -vv
```
https://github.com/ovn-org/ovn-kubernetes/actions/runs/6646915605/job/18062050736?pr=3979

**Comment by npinaeva** [comment]
one more https://github.com/ovn-org/ovn-kubernetes/actions/runs/6734120950/job/18305537344?pr=3995

**Comment by flavio-fernandes** [comment]
Flake seen here:  https://github.com/ovn-org/ovn-kubernetes/pull/4050  ( https://github.com/ovn-org/ovn-kubernetes/actions/runs/7199631620/job/19612611460?pr=4050 )

**Comment by tssurya** [comment]
one more here: https://github.com/ovn-org/ovn-kubernetes/actions/runs/7789676840/job/21242664047?pr=4100 :/ 

https://github.com/ovn-org/ovn-kubernetes/pull/4100

@maiqueb or @qinqon : can one of you please fix this?

**Comment by qinqon** [comment]
> one more here: https://github.com/ovn-org/ovn-kubernetes/actions/runs/7789676840/job/21242664047?pr=4100 :/
> 
> #4100
> 
> @maiqueb or @qinqon : can one of you please fix this?

@tssurya we are preparing a fix for it, it will be ready to merge today.

**Comment by tssurya** [comment]
one more: https://github.com/ovn-org/ovn-kubernetes/actions/runs/7813480320/job/21313563225?pr=4100#logs
ack thanks @martinkennelly !!

**Comment by qinqon** [comment]
@tssurya we are going to start small https://github.com/ovn-org/ovn-kubernetes/pull/4140

We have some other ideas but we will go step by step.

**Comment by martinkennelly** [comment]
Thanks @qinqon 

seen again here: https://github.com/ovn-org/ovn-kubernetes/actions/runs/7814875289/job/21318048354?pr=4061

**Comment by qinqon** [comment]
> Thanks @qinqon
> 
> seen again here: https://github.com/ovn-org/ovn-kubernetes/actions/runs/7814875289/job/21318048354?pr=4061

Adding error so I can track if it's always the same.

```
 [FAILED] Timed out after 60.001s.
  worker1: after live migration to node owning the subnet: Check tcp connection is not broken
  Expected success, but got an error:
      <*errors.errorString | 0xc000f52b70>: 
      http connection to virtual machine was broken
      {
          s: "http connection to virtual machine was broken",
      }
  In [It] at: /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/kubevirt.go:330 @ 02/07/24 13:22:07.982

```

**Comment by qinqon** [comment]
I am preoducing this error at my fork running 12 jobs in parallel and a more simple tcp server so is easier to debug
https://github.com/qinqon/ovn-kubernetes/pull/10

https://github.com/qinqon/ovn-kubernetes/pull/10#issuecomment-1934147363

**Comment by qinqon** [comment]
After around 25 runs show this failure with the stabilization PR
https://github.com/ovn-org/ovn-kubernetes/actions/runs/7840556162/job/21395793644?pr=4145
```
2024-02-09T07:20:48.1135328Z   Latency metrics for node ovn-worker3
2024-02-09T07:20:48.1136685Z   [1mSTEP:[0m Destroying namespace "kv-live-migration-923" for this suite. [38;5;243m@ 02/09/24 07:20:48.113[0m
2024-02-09T07:20:48.1181879Z [38;5;9m• [FAILED] [419.375 seconds][0m
2024-02-09T07:20:48.1184612Z [0mKubevirt Virtual Machines [38;5;243mwhen live migration [38;5;9m[1m[It] with pre-copy succeeds, should keep connectivity[0m
2024-02-09T07:20:48.1187469Z [38;5;243m/home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/kubevirt.go:836[0m
2024-02-09T07:20:48.1188240Z 
2024-02-09T07:20:48.1190075Z   [38;5;9m[FAILED] worker1: after live migration for the second time to node not owning subnet: Check connectivity is restored after delete deny all network policy
2024-02-09T07:20:48.1191653Z   Expected success, but got an error:
2024-02-09T07:20:48.1192369Z       <*fmt.wrapError | 0xc000763d20>: 
2024-02-09T07:20:48.1193746Z       failed Write to server: write tcp 172.18.0.1:41446->172.18.0.3:31702: write: broken pipe
2024-02-09T07:20:48.1194723Z       {
2024-02-09T07:20:48.1196036Z           msg: "failed Write to server: write tcp 172.18.0.1:41446->172.18.0.3:31702: write: broken pipe",
2024-02-09T07:20:48.1197218Z           err: <*net.OpError | 0xc00123f9f0>{
2024-02-09T07:20:48.1197908Z               Op: "write",
2024-02-09T07:20:48.1198471Z               Net: "tcp",
2024-02-09T07:20:48.1199748Z               Source: <*net.TCPAddr | 0xc0006ed9b0>{IP: [172, 18, 0, 1], Port: 41446, Zone: ""},
2024-02-09T07:20:48.1201354Z               Addr: <*net.TCPAddr | 0xc0006edaa0>{IP: [172, 18, 0, 3], Port: 31702, Zone: ""},
2024-02-09T07:20:48.1202601Z               Err: <*os.SyscallError | 0xc000763d00>{
2024-02-09T07:20:48.1203335Z                   Syscall: "write",
2024-02-09T07:20:48.1203895Z                   Err: <syscall.Errno>0x20,
2024-02-09T07:20:48.1204307Z               },
2024-02-09T07:20:48.1204642Z           },
2024-02-09T07:20:48.1204937Z       }[0m
2024-02-09T07:20:48.1206276Z   [38;5;9mIn [1m[It][0m[38;5;9m at: [1m/home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/kubevirt.go:401[0m [38;5;243m@ 02/09/24 07:20:46.845[0m
```

I will try to reproduce this too with the testing PR.

**Comment by qinqon** [comment]
I have disable interconnect and remove the policy part of the test to make it simpler also printing all the echos
this is the result
Server logs
```
Server is running on: :9900
  2024/02/12 12:26:15 Handling connection 100.64.0.5:32870
  2024/02/12 12:26:15 Handling connection [fd98::5]:58614
  2024/02/12 12:28:33 failed copying data: readfrom tcp 10.244.1.8:9900->100.64.0.5:32870: splice: connection reset by peer
  2024/02/12 12:28:33 Closing connection 100.64.0.5:32870
```
Client logs
```
STEP: worker1: after live migration for the second time to node not owning subnet: Check tcp connection is not broken @ 02/12/24 13:07:39.455
  STEP: Writing 'Halo' @ 02/12/24 13:07:39.455
  STEP: Reading 'Halo' @ 02/12/24 13:07:39.455
  STEP: failed reading: read tcp 172.18.0.1:41882->172.18.0.4:30247: read: connection reset by peer  @ 02/12/24 13:07:39.465
  STEP: Writing 'Halo' @ 02/12/24 13:07:54.465
  STEP: failed writing: write tcp 172.18.0.1:41882->172.18.0.4:30247: write: broken pipe  @ 02/12/24 13:07:54.466
  STEP: Writing 'Halo' @ 02/12/24 13:08:09.466
  STEP: failed writing: write tcp 172.18.0.1:41882->172.18.0.4:30247: write: broken pipe  @ 02/12/24 13:08:09.467
  STEP: Writing 'Halo' @ 02/12/24 13:08:24.467
  STEP: failed writing: write tcp 172.18.0.1:41882->172.18.0.4:30247: write: broken pipe  @ 02/12/24 13:08:24.467
```

I am going to attach a tcpdump to the client to see if we are receving at RST packet that triggers the "connection reset by peer" at first echo.

**Comment by qinqon** [comment]
This should appear way less often after 
https://github.com/ovn-org/ovn-kubernetes/pull/4174

**Comment by npinaeva** [comment]
https://github.com/ovn-org/ovn-kubernetes/actions/runs/8449224256/job/23144057157?pr=4154

**Comment by trozet** [comment]
https://github.com/ovn-org/ovn-kubernetes/actions/runs/8478907825/job/23232584804?pr=4246

**Comment by qinqon** [comment]
> https://github.com/ovn-org/ovn-kubernetes/actions/runs/8478907825/job/23232584804?pr=4246

@trozet adding this link to https://github.com/ovn-org/ovn-kubernetes/issues/4237 since is the specific error related to nameserver

**Comment by martinkennelly** [comment]
https://github.com/ovn-org/ovn-kubernetes/actions/runs/8648860156/job/23715049496?pr=4271

Seen in https://github.com/ovn-org/ovn-kubernetes/pull/4271

**Comment by flavio-fernandes** [comment]
Also seen in https://github.com/ovn-org/ovn-kubernetes/pull/4468

**Comment by martinkennelly** [comment]
Seen again

**Comment by oshoval** [comment]
bit different because it is primary but also flakes and is "with pre-copy succeeds, should keep connectivity"

Kubevirt Virtual Machines with default pod network when live migration [It] with pre-copy succeeds, should keep connectivity

https://github.com/ovn-org/ovn-kubernetes/actions/runs/10938484527/job/30367928920?pr=4708

if needed will open new issue for it

**Comment by flavio-fernandes** [comment]
Seen again here: https://github.com/ovn-org/ovn-kubernetes/actions/runs/11449422218/job/31855967593?pr=4508

**Comment by ormergi** [comment]
https://github.com/ovn-org/ovn-kubernetes/pull/4612

**Comment by flavio-fernandes** [comment]
Seen again here: https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/14386806890/job/40346051511?pr=5173

**Comment by npinaeva** [comment]
@qinqon I see there is a linked PR https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4416 that got stale. What is the plan for this issue?

**Comment by qinqon** [comment]
> [@qinqon](https://github.com/qinqon) I see there is a linked PR [#4416](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4416) that got stale. What is the plan for this issue?

Iam backporting that old PoC to see if we can productify it and also use iperf3 to test live migration like the other test so we don't have different way to test for default pod network migration, it also improve live migration for east/west tcp connections.

**Comment by qinqon** [comment]
Note: #6044 tracks the **same test** (`with pre-copy succeeds, should keep connectivity`) but a **different failure mode**:

- **Here (#3986):** TCP connection was established but **breaks during/after migration** — a network path reconvergence problem (`"http connection to virtual machine was broken"`).
- **#6044:** The `tcpserver` podman container inside the guest **is missing entirely** after migration (`"no container with name or ID tcpserver found"`) — a guest-internal VM lifecycle/setup problem.

Updated title to distinguish them.

**Comment by qinqon** [comment]
PR #6637 ("e2e(kubevirt): replace echoserver with iperf3") replaces the custom golang echoserver and TCP dial/echo/reconnect plumbing with iperf3. This eliminates the fragile HTTP connection check that caused this flake (`"http connection to virtual machine was broken"`).

Linking as the fix. See also #6044 (same test, different failure mode — tcpserver container missing).

**Comment by qinqon** [comment]
Closing as obsolete. PR #6637 (merged Jul 7) replaced the echoserver-based connectivity check with iperf3 for this test, so the exact echoserver/TCP-broken code path this issue tracked no longer exists.

The underlying test ("with pre-copy succeeds, should keep connectivity", default pod network) is still flaky post-#6637, but with a new iperf3-specific symptom — tracked separately in #6684 (IPv6 iperf3 traffic fails to recover after migration).
